### PR TITLE
Move node-cli from keytar to crypto and node-persist for refresh tokens

### LINKED ISSUE #192: Replace Keytar (node-cli)
Options for replacing keytar for non-electron applications are less clear.

### PR DESCRIPTION
**Pull request by hl662**
- use node-persist for persistent storage of refresh token
- use Node's built in crypto module to encrypt and decrypt refresh tokens stored
- add option to configure refresh token storage path
- add unit tests for the token store
Closes #192

### CODE REVIEW

File: packages/node-cli/src/TokenStore.ts [hunk 1]
```diff
@@ -3,9 +3,11 @@
 * See LICENSE.md in the project root for license terms and full copyright notice.
 *--------------------------------------------------------------------------------------------*/
 
-import { getPassword, setPassword } from "keytar";
 import * as OperatingSystemUserName from "username";
 import { TokenResponse } from "@openid/appauth";
+import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from "crypto";
```
Comment (aruniverse) [inline packages/node-cli/src/TokenStore.ts:8]:
```suggestion
import { createCipheriv, createDecipheriv, randomBytes, scryptSync } from "node:crypto";
```
for consistency

File: packages/node-cli/src/TokenStore.ts [hunk 2]
```diff
@@ -14,37 +16,84 @@ import { TokenResponse } from "@openid/appauth";
...
-
-  public constructor(namedArgs: {clientId: string, issuerUrl: string, scopes: string}) {
+  private readonly _store: NodePersist.LocalStorage;
+  public constructor(namedArgs: {clientId: string, issuerUrl: string, scopes: string}, dir?: string) {
     // A stored credential is only valid for a combination of the clientId, the issuing authority and the requested scopes.
     // We make the storage key a combination of clientId and issuing authority so that keys can stay cached when switching
     // between PROD and QA environments.
     // We store the scopes in our password blob so we know if a new token is required due to updated scopes.
-    this._appStorageKey = `iTwinJs:${namedArgs.clientId}:${namedArgs.issuerUrl}`;
+    const configFileName = `itwinJs_${namedArgs.clientId}`;
```
Comment (aruniverse) [inline packages/node-cli/src/TokenStore.ts:25]:
```suggestion
    const configFileName = `iTwinJs_${namedArgs.clientId}`;
```

is this the same structure as the electron auth token?

File: common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@itwin/node-cli-authorization",
+      "comment": "Move node-cli from keytar to use crypto and node-persist",
+      "type": "minor"
```
Comment (aruniverse) [inline common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json:6]: well this is breaking

**Review by aruniverse — COMMENTED** [review]
how do we handle when new scopes are added to a client and they need to resign in?

**Comment by hl662** [comment]
> how do we handle when new scopes are added to a client and they need to resign in?

The TokenStore compares the new scopes with the scopes in the cached tokenResponse. If different, the load() returns undefined, prompting the user to sign in again. For extra measures, I'll delete the cache entry as well if scopes are mismatched, in case user doesn't go through the resign process (also added a unit test to confirm this behavior)

File: common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json (hunk 3 above)
Comment (hl662) [reply common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json]: changed type to `major`

File: common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json
Comment (aruniverse) [inline common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json:1]: can delete this file

Comment (hl662) [reply common/changes/@itwin/node-cli-authorization/nam-node-cli-keytar_2023-08-21-15-16.json:1]: done

**Review by ben-polinsky — APPROVED** [review]

File: packages/node-cli/src/TokenStore.ts [hunk 4]
```diff
@@ -14,39 +16,88 @@ import { TokenResponse } from "@openid/appauth";
...
+   * @returns
+   */
+  private generateCipherKey(): Buffer {
+    return scryptSync(this._appStorageKey, "iTwin", 32); // aes-256-cbc requires a key length of 32 bytes.
+  }
+  /**
+   * Uses node's native `crypto` module to encrypt the given cache entry.
+   * @returns an object containing a hexadecimal encoded token, returned as a string, as well as the initialization vector.
+   */
+  private encryptCache(cacheEntry: object): {encryptedCache: string, iv: string} {
```
Comment (ben-polinsky) [inline packages/node-cli/src/TokenStore.ts:61]: Can this cacheEntry be better typed?

**Comment by GintV** [comment]
> > how do we handle when new scopes are added to a client and they need to resign in?
> 
> The TokenStore compares the new scopes with the scopes in the cached tokenResponse. If different, the load() returns undefined, prompting the user to sign in again. For extra measures, I'll delete the cache entry as well if scopes are mismatched, in case user doesn't go through the resign process (also added a unit test to confirm this behavior)

So does this mean that clients don't need to check accessToken if it has all the scopes that we require? Atm we do that since in the past we had an issue with this

**Comment by hl662** [comment]
> > > how do we handle when new scopes are added to a client and they need to resign in?
> > 
> > 
> > The TokenStore compares the new scopes with the scopes in the cached tokenResponse. If different, the load() returns undefined, prompting the user to sign in again. For extra measures, I'll delete the cache entry as well if scopes are mismatched, in case user doesn't go through the resign process (also added a unit test to confirm this behavior)
> 
> So does this mean that clients don't need to check accessToken if it has all the scopes that we require? Atm we do that since in the past we had an issue with this
> 
> cc @DenisMicha

I'm not too sure with that, what the check does is compare if the 2 scopes are different, not necessarily if you have all the scopes your application needs for its purposes). Lastly, this comparison check only exists for the node-cli authentication clients, as far as I can tell.

**Pull request merged**